### PR TITLE
Add `proxy_to_pthread` option to `platform=web`

### PR DESCRIPTION
**Pull request by adamscott**
This PR adds the option `proxy_to_pthread` to the `web` platform (with `proxy_to_pthread=yes` as default value).

The emscripten library compiles C/C++ apps to WASM. Unfortunately, WASM, when ran in the main thread, runs synchronously, making it block the main thread. That makes it so that if the application is in a computing intensive task, it will block the page main thread, affecting the UI and the usability of that page.

When compiled with this PR, users can use the web page (like entering text in a form) without any noticeable issues. Currently, it is impossible to write such text while the engine is loading. [See that Firefox profiler as an example.](https://profiler.firefox.com/public/jvac57xxa4cfqaazyxshq37w0g8q7xqmc6jvc9g/calltree/?globalTrackOrder=d0wc&hiddenGlobalTracks=1w79wb&hiddenLocalTracksByPid=203065-1w4~212678-0~212002-01~203236-0w3~212359-0wbdwg&thread=d&v=9)

| image |
| :---: |
| See the red line in the Isolate Web Content row, user input is impossible during that time |

[Emscripten offers a linker option to run the application in a web worker](https://emscripten.org/docs/porting/pthreads.html#additional-flags), effectively fixing the issue mentioned above. Web workers are the equivalent of threads for the web platform.

With my PR, if the user add the `do_no_block_main_thread=yes` option to it's `platform=web` scons build, the builder adds `-s PROXY_TO_PTHREAD` to the linker options to create a stub `main()` function that will act as a proxy to the newly created main web worker.

> **Note** 
> My PR will add the `-s TEXTDECODER=0` option to the linker too, as there was [non-consistent issues about internal emscripten calls](https://github.com/emscripten-core/emscripten/issues/18034). The suggested solution from the developer seems to deactivate `TEXTDECODER`, as [there no big impact on the usability of the application](https://github.com/emscripten-core/emscripten/issues/18034#issuecomment-1277561925).

[The Javascript C API](https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html#implement-a-c-api-in-javascript) ("platform/web/js/libs") had to be updated to specify the `__proxy` setting (now set to `sync`). This is because otherwise, the web worker will use it's default `Module` values that were initialized with the web worker. But all the data set before the initialization still reside in the main thread.

Setting the value `sync` to each function `__proxy` option makes it so that emscripten knows that the C API Javascript function must be run on the main thread instead of in the main web worker.

---

Here's a demo. Don't mind the non-fluid mouse movement, it's related to the screen capture, it was super smooth while recording. But notice how the HTML UI react even if either the editor blocks the thread when loading (as usual) or while the game blocks the thread each frame for a whole second.

[Capture vidéo du 2023-07-29 14-31-18.webm](https://github.com/godotengine/godot/assets/270928/ad53296c-9015-486f-97d8-6457ba937575)

---

**TL;DR** Your game will still freeze under heavy processing on the main thread. But, the web page hosting the game will not freeze, as the Godot game will run in it's own thread instead of running on the web page's main thread.

---

> **Note**
> ~~This PR needs testing. There's some issues known issues about this PR. (when fixed, this PR will drop it's draft status.)~~
> - ~~input is broken.~~ (fixed)

### CODE REVIEW

**Comment by Faless** [comment]
Woah! That's really amazing :blue_heart:  , I've been wanting to do this for a looooooong time!
It'll need some testing, but it's looking great :trophy: !

**Comment by dsnopek** [comment]
Thanks so much for taking this on!

> Setting the value `sync` to each function `__proxy` option makes it so that emscripten knows that the C API Javascript function must be run on the main thread instead of in the main web worker.

There's some Javascript code elsewhere in Godot's code base, for example in these modules:

- websocket
- webrtc
- webxr

If you haven't checked already, it'd be good to double check that they don't also need `__proxy` added.

**Comment by adamscott** [comment]
> There's some Javascript code elsewhere in Godot's code base, for example in these modules:
>  - websocket
>  - webrtc
>  - webxr

Websocket and WebRTC changes are already in my PR. 

And for WebXR, a certain @dsnopek already added the `__proxy` prop to `sync`, 3 years ago. In fact, I should thank you for having me realize that the `__proxy` setting vaguely mentioned in the emscripten docs was to set the value as you did. :upside_down_face:

**Comment by dsnopek** [comment]
> Websocket and WebRTC changes are already in my PR.

Ack, sorry, not sure how I glossed over that :-)

File: platform/web/js/libs/library_godot_fetch.js [hunk 1]
```diff
@@ -218,6 +223,17 @@ const GodotFetch = {
...
+	godot_js_fetch_body_length_get__proxy: 'sync',
+	godot_js_fetch_body_length_get__sig: 'ii',
+	godot_js_fetch_body_length_get: function (p_id) {
+		const obj = IDHandler.get(p_id);
+		if (!obj || !obj.response) {
+			return -1;
+		}
+		return obj.bodySize;
+	},
+
```
Comment (Faless) [inline platform/web/js/libs/library_godot_fetch.js:235]: This function was removed in #79846

**Comment by adamscott** [comment]
I added a demo in the PR description that shows my PR load successfully the editor and runs a game without blocking the main Javascript thread.

File: platform/web/detect.py [hunk 2]
```diff
@@ -213,6 +218,14 @@ def configure(env: "Environment"):
         env.Append(LINKFLAGS=["-s", "SIDE_MODULE=2"])
         env.extra_suffix = ".dlink" + env.extra_suffix
 
+    # Run the main application in a web worker
+    if env["do_not_block_main_thread"]:
+        env.Append(LINKFLAGS=["-s", "PROXY_TO_PTHREAD=1"])
+        env.Append(CPPDEFINES=["PROXY_TO_PTHREAD_ENABLED"])
+        env.Append(LINKFLAGS=["-s", "EXPORTED_RUNTIME_METHODS=['_emscripten_proxy_main']"])
```
Comment (adamscott) [inline platform/web/detect.py:225]: I added this exported runtime method here because otherwise, emscripten would complain that it is not exposed. It seems possible to append multiple `-s EXPORTED_RUNTIME_METHODS` parameters, as I did not have problems running the editor or a game, see my demo.

File: platform/web/js/libs/library_godot_fetch.js (hunk 1 above)
Comment (adamscott) [reply platform/web/js/libs/library_godot_fetch.js]: Removed!

**Comment by akien-mga** [comment]
That looks really cool, thanks for working on it!

Some minor comments:
- `do_not_block_main_thread` is a mouthful, we don't really have options which are that verbose / full sentences like this. It's not a big deal but maybe `proxy_to_pthread` would be more fitting, and matches the Emscripten option and the `PROXY_TO_PTHREAD_ENABLED` define? It's also long-ish, but 3 words vs 5 is nice ;)
- Are there specific concerns to make this the default behavior, i.e. opt-out instead of opt-in?

**Comment by Faless** [comment]
Once again, this is awesome :heart_eyes_cat: .
I have been testing this, and found few issues though:

With `dev_build=yes`  (tested on Firefox 116)
---

In the editor, as soon as the mouse enter the window, with the console open, I'm getting these errors (tested on Firefox):

```
Window::_event_callback(DisplayServer::WindowEvent)@:wasm-function[69315]:0x17d344b
void call_with_variant_args_helper<Window, DisplayServer::WindowEvent, 0ul>(Window*, void (Window::*)(DisplayServer::WindowEvent), Variant const**, Callable::CallError&, IndexSequence<0ul>)@:wasm-function[69585]:0x17f336d
CallableCustomMethodPointer<Window, DisplayServer::WindowEvent>::call(Variant const**, int, Variant&, Callable::CallError&) const@:wasm-function[69584]:0x17f31c3
Callable::callp(Variant const**, int, Variant&, Callable::CallError&) const@:wasm-function[143572]:0x2cfaf09
DisplayServerWeb::send_window_event_callback(int)@:wasm-function[506]:0x76df9
void call_with_variant_args_static<int, 0ul>(void (*)(int), Variant const**, Callable::CallError&, IndexSequence<0ul>)@:wasm-function[997]:0x79c10
CallableCustomStaticMethodPointerRet<void, int>::call(Variant const**, int, Variant&, Callable::CallError&) const@:wasm-function[996]:0x79bd0
Callable::callp(Variant const**, int, Variant&, Callable::CallError&) const@:wasm-function[143572]:0x2cfaf09
CallableCustomBind::call(Variant const**, int, Variant&, Callable::CallError&) const@:wasm-function[143633]:0x2cfd5da
Callable::callp(Variant const**, int, Variant&, Callable::CallError&) const@:wasm-function[143572]:0x2cfaf09
CallQueue::_call_function(Callable const&, Variant const*, int, bool)@:wasm-function[153631]:0x2f08992
CallQueue::flush()@:wasm-function[153633]:0x2f0920b
SceneTree::physics_process(double)@:wasm-function[67938]:0x176a14c
Main::iteration()@:wasm-function[1388]:0x9b7ee
OS_Web::main_loop_iterate()@:wasm-function[1177]:0x7ee5b
main_loop_callback()@:wasm-function[1166]:0x7e98d
godot_web_main(int, char**)@:wasm-function[1167]:0x7ec4c
main@:wasm-function[1283]:0x83844
__main_void@:wasm-function[160463]:0x30f7f4f
_main_thread@:wasm-function[160465]:0x30f801c
invokeEntryPoint@http://127.0.0.1:8060/godot.editor.js:3837:37
handleMessage@http://127.0.0.1:8060/godot.editor.worker.js:124:16
godot.editor.worker.js:24:11
    threadPrintErr http://127.0.0.1:8060/godot.editor.worker.js:24
    handleMessage http://127.0.0.1:8060/godot.editor.worker.js:152
```

With regular builds (i.e. `dev_build=no`)  (tested on Firefox 116, and Chromium 115)
---

Opening the console, and closing it, generates the following artifacts:

corrupted_size

And further resizing the window then seems to not fix the issue.

Also, and more importantly, whenever in the editor, I try to create a Mesh (to assign to a MeshInstance3D), I'm getting the following error (and the editor crashes):

```
Pthread 0x013e8db8 sent an error! http://127.0.0.1:8060/godot.editor.js:8677: Uncaught TypeError: GLctx.getBufferSubData is not a function
onPrintError @ godot.editor.js:13776
worker.onerror @ godot.editor.js:3713
godot.editor.js:3714 Uncaught ErrorEvent {isTrusted: true, message: 'Uncaught TypeError: GLctx.getBufferSubData is not a function', filename: 'http://127.0.0.1:8060/godot.editor.js', lineno: 8677, colno: 16, …}
worker.onerror @ godot.editor.js:3714
godot.editor.js:8677 Uncaught TypeError: GLctx.getBufferSubData is not a function
    at _glGetBufferSubData (godot.editor.js:8677:16)
    at GLES3::Utilities::buffer_get_data(unsigned int, unsigned int, unsigned int) (15a0ca8a:0xb28ef6)
    at GLES3::MeshStorage::mesh_get_surface(RID, int) const (15a0ca8a:0xaf3186)
    at RenderingServerDefault::mesh_get_surface(RID, int) const (15a0ca8a:0x2a8e463)
    at RenderingServer::mesh_surface_get_arrays(RID, int) const (15a0ca8a:0x26c3d26)
    at PrimitiveMesh::surface_get_arrays(int) const (15a0ca8a:0x22ae5e5)
    at Mesh::generate_triangle_mesh() const (15a0ca8a:0x2211391)
    at MeshInstance3DGizmoPlugin::redraw(EditorNode3DGizmo*) (15a0ca8a:0x173954b)
    at EditorNode3DGizmo::redraw() (15a0ca8a:0x145e10e)
    at Node3D::_update_gizmos() (15a0ca8a:0x1d6943d)
```

This might be an emscripten bug (i.e. `glGetBufferSubData` not being proxied to main).

**Comment by Faless** [comment]
> This might be an emscripten bug (i.e. `glGetBufferSubData` not being proxied to main).

Indeed, using our own proxied implementation of `glGetBufferSubData` seems to fix this issue:

```diff
diff --git a/drivers/gles3/storage/particles_storage.cpp b/drivers/gles3/storage/particles_storage.cpp
index b7ebe055f5..fe7fcb8c11 100644
--- a/drivers/gles3/storage/particles_storage.cpp
+++ b/drivers/gles3/storage/particles_storage.cpp
@@ -775,7 +775,7 @@ void ParticlesStorage::particles_set_view_axis(RID p_particles, const Vector3 &p
 		LocalVector<ParticleInstanceData3D> particle_vector;
 		particle_vector.resize(particles->amount);
 		particle_array = particle_vector.ptr();
-		glGetBufferSubData(GL_ARRAY_BUFFER, 0, particles->amount * sizeof(ParticleInstanceData3D), particle_array);
+		godot_webgl2_glGetBufferSubData(GL_ARRAY_BUFFER, 0, particles->amount * sizeof(ParticleInstanceData3D), particle_array);
 #endif
 		SortArray<ParticleInstanceData3D, ParticlesViewSort> sorter;
 		sorter.compare.z_dir = axis;
@@ -1133,7 +1133,7 @@ void ParticlesStorage::_particles_reverse_lifetime_sort(Particles *particles) {
 	LocalVector<ParticleInstanceData> particle_vector;
 	particle_vector.resize(particles->amount);
 	particle_array = particle_vector.ptr();
-	glGetBufferSubData(GL_ARRAY_BUFFER, 0, buffer_size, particle_array);
+	godot_webgl2_glGetBufferSubData(GL_ARRAY_BUFFER, 0, buffer_size, particle_array);
 #endif
 
 	uint32_t lifetime_split = (MIN(int(particles->amount * particles->sort_buffer_phase), particles->amount - 1) + 1) % particles->amount;
diff --git a/drivers/gles3/storage/utilities.cpp b/drivers/gles3/storage/utilities.cpp
index 3e6e72edad..29542b0407 100644
--- a/drivers/gles3/storage/utilities.cpp
+++ b/drivers/gles3/storage/utilities.cpp
@@ -113,7 +113,7 @@ Vector<uint8_t> Utilities::buffer_get_data(GLenum p_target, GLuint p_buffer, uin
 #if defined(__EMSCRIPTEN__)
 	{
 		uint8_t *w = ret.ptrw();
-		glGetBufferSubData(p_target, 0, p_buffer_size, w);
+		godot_webgl2_glGetBufferSubData(p_target, 0, p_buffer_size, w);
 	}
 #else
 	void *data = glMapBufferRange(p_target, 0, p_buffer_size, GL_MAP_READ_BIT);
diff --git a/platform/web/godot_webgl2.h b/platform/web/godot_webgl2.h
index d2f46e125e..3ade9e4239 100644
--- a/platform/web/godot_webgl2.h
+++ b/platform/web/godot_webgl2.h
@@ -44,6 +44,7 @@ extern "C" {
 #endif
 
 void godot_webgl2_glFramebufferTextureMultiviewOVR(GLenum target, GLenum attachment, GLuint texture, GLint level, GLint baseViewIndex, GLsizei numViews);
+void godot_webgl2_glGetBufferSubData(GLenum target, GLintptr offset, GLsizeiptr size, GLvoid *data);
 
 #define glFramebufferTextureMultiviewOVR godot_webgl2_glFramebufferTextureMultiviewOVR
 
diff --git a/platform/web/js/libs/library_godot_webgl2.js b/platform/web/js/libs/library_godot_webgl2.js
index 3c6de4a071..dbaec9f01b 100644
--- a/platform/web/js/libs/library_godot_webgl2.js
+++ b/platform/web/js/libs/library_godot_webgl2.js
@@ -32,6 +32,19 @@ const GodotWebGL2 = {
 	$GodotWebGL2__deps: ['$GL', '$GodotRuntime'],
 	$GodotWebGL2: {},
 
+	// This is implemented as "glGetBufferSubData" in new emscripten versions.
+	// Since we have to support older (pre 2.0.17) emscripten versions, we add this wrapper function instead.
+	godot_webgl2_glGetBufferSubData__proxy: 'sync',
+	godot_webgl2_glGetBufferSubData__sig: 'vippp',
+	godot_webgl2_glGetBufferSubData__deps: ['$GL', 'emscripten_webgl_get_current_context'],
+	godot_webgl2_glGetBufferSubData: function (target, offset, size, data) {
+		const gl_context_handle = _emscripten_webgl_get_current_context(); // eslint-disable-line no-undef
+		const gl = GL.getContext(gl_context_handle);
+		if (gl) {
+			gl.GLctx['getBufferSubData'](target, offset, HEAPU8, data, size);
+		}
+	},
+
 	godot_webgl2_glFramebufferTextureMultiviewOVR__deps: ['emscripten_webgl_get_current_context'],
 	godot_webgl2_glFramebufferTextureMultiviewOVR__proxy: 'sync',
 	godot_webgl2_glFramebufferTextureMultiviewOVR__sig: 'viiiiii',
```

Though we should probably try to fix this upstream.

**Comment by Faless** [comment]
The functions in upstream [`library_webgl2.js`](https://github.com/emscripten-core/emscripten/blob/main/src/library_webgl2.js) doesn't seem to have the `__proxy` annotation, adding `glGetBufferSubData__proxy: 'sync'` to the upstream library seems indeed to fix the issue.

**Comment by sakrel** [comment]
I'm surprised there's a need to proxy any gl calls at all. AFAIK chrome and Firefox have support for the [OffscreenCanvas](https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas) API which should make it possible to execute gl calls on the worker thread.

**Comment by Faless** [comment]
OffscreenCanvas is [not yet supported by Safari](https://caniuse.com/offscreencanvas)

**Comment by Faless** [comment]
For reference, I think there is an upstream emscripten issue about that: https://github.com/emscripten-core/emscripten/issues/19754

**Comment by adamscott** [comment]
@akien-mga Sorry, just read your comment.

> Some minor comments:
> 
>     * `do_not_block_main_thread` is a mouthful, we don't really have options which are that verbose / full sentences like this. It's not a big deal but maybe `proxy_to_pthread` would be more fitting, and matches the Emscripten option and the `PROXY_TO_PTHREAD_ENABLED` define? It's also long-ish, but 3 words vs 5 is nice ;)

Yes, good idea. Will change to `proxy_to_pthread`. I thought of it at first, but as it does more than just setting `-DPROXY_TO_PTHREAD_ENABLED`, I changed it to `do_not_block_main_thread`.

>     * Are there specific concerns to make this the default behavior, i.e. opt-out instead of opt-in?

In my mind, this should be a opt-out feature rather than the current opt-in. But I'm worried about certain side-effects that could occur. But I'm all in to change my PR to make it opt-out, as it greatly improves the quality of life of the user.

**Comment by adamscott** [comment]
> Opening the console, and closing it, generates the following artifacts:
> 
> corrupted_size

Found a fix for that issue! I fixed it by using `emscripten_set_canvas_element_size()` to tell the framework to update it's internal rendering viewport. It seems that this issue don't appear when Godot runs on the main thread.

File: platform/web/os_web.h [hunk 3]
```diff
@@ -88,9 +88,12 @@ class OS_Web : public OS_Unix {
 	String get_executable_path() const override;
 	Error shell_open(String p_uri) override;
 	String get_name() const override;
+
+#ifndef PROXY_TO_PTHREAD_ENABLED
```
Comment (Faless) [inline platform/web/os_web.h:92]: I'm not very fond of ifdefs in header files, can we move this to the CPP file instead?
```c++
void OS_Web::add_frame_delay(bool p_can_draw) {
#ifndef PROXY_TO_PTHREAD_ENABLED
        OS_Unix::add_frame_delay(p_can_draw);
#endif
}
```

File: platform/web/display_server_web.h [hunk 4]
```diff
@@ -88,14 +88,18 @@ class DisplayServerWeb : public DisplayServer {
...
 	static void touch_callback(int p_type, int p_count);
-	static void key_callback(int p_pressed, int p_repeat, int p_modifiers);
-	static void vk_input_text_callback(const char *p_text, int p_cursor);
-	static void gamepad_callback(int p_index, int p_connected, const char *p_id, const char *p_guid);
+	static void _key_callback(int p_pressed, int p_repeat, int p_modifiers);
+	static void key_callback(String p_key_event_code, String p_key_event_key, int p_pressed, int p_repeat, int p_modifiers);
+	static void _vk_input_text_callback(const char *p_text, int p_cursor);
+	static void vk_input_text_callback(const String p_text, int p_cursor);
+	static void _gamepad_callback(int p_index, int p_connected, const char *p_id, const char *p_guid);
+	static void gamepad_callback(int p_index, int p_connected, const String p_id, const String p_guid);
```
Comment (AThousandShips) [inline platform/web/display_server_web.h:102]:
```suggestion
	static void gamepad_callback(int p_index, int p_connected, const String &p_id, const String &p_guid);
```

File: platform/web/display_server_web.h [hunk 5]
```diff
@@ -88,14 +88,18 @@ class DisplayServerWeb : public DisplayServer {
...
 	static void mouse_move_callback(double p_x, double p_y, double p_rel_x, double p_rel_y, int p_modifiers);
 	static int mouse_wheel_callback(double p_delta_x, double p_delta_y);
 	static void touch_callback(int p_type, int p_count);
-	static void key_callback(int p_pressed, int p_repeat, int p_modifiers);
-	static void vk_input_text_callback(const char *p_text, int p_cursor);
-	static void gamepad_callback(int p_index, int p_connected, const char *p_id, const char *p_guid);
+	static void _key_callback(int p_pressed, int p_repeat, int p_modifiers);
+	static void key_callback(String p_key_event_code, String p_key_event_key, int p_pressed, int p_repeat, int p_modifiers);
+	static void _vk_input_text_callback(const char *p_text, int p_cursor);
+	static void vk_input_text_callback(const String p_text, int p_cursor);
```
Comment (AThousandShips) [inline platform/web/display_server_web.h:100]:
```suggestion
	static void vk_input_text_callback(const String &p_text, int p_cursor);
```

File: platform/web/os_web.h (hunk 3 above)
Comment (adamscott) [reply platform/web/os_web.h]: Done!

**Review by Faless — APPROVED** [review]
Amazing work :rocket: :1st_place_medal: ! 

It seems that dlink + proxy_to_pthread is broken (needs further investigation), I suggest disabling it in that case but leave it on by default:

```diff
diff --git a/platform/web/detect.py b/platform/web/detect.py
index 4ad66d0b25..d306869889 100644
--- a/platform/web/detect.py
+++ b/platform/web/detect.py
@@ -216,6 +216,10 @@ def configure(env: "Environment"):
             env.Append(LINKFLAGS=["-Wl,-u,scalbnf"])
 
     if env["dlink_enabled"]:
+        if env["proxy_to_pthread"]:
+            print("GDExtension support requires proxy_to_pthread=no, disabling")
+            env["proxy_to_pthread"] = False
+
         if cc_semver < (3, 1, 14):
             print("GDExtension support requires emscripten >= 3.1.14, detected: %s.%s.%s" % cc_semver)
             sys.exit(255)
```

I'm approving because all my other tests (including testing the 2d and 3d platformer demos) worked well.

**Comment by akien-mga** [comment]
Thanks!

**Comment by bballeysark** [comment]
When I build my web game project with `platform=web threads=no`, it seems that the `add_frame_delay` calls `nanosleep` and it takes unnecessary time to sleep to match max_fps. From the os_web.h comment says that calling delay_usec (which is calling nanosleep) blocks the main thread so it override it to do nothing.
```
// Override default OS implementation which would block the main thread with delay_usec.
// Implemented in web_main.cpp loop callback instead.
void add_frame_delay(bool p_can_draw) override;
```

I think it can affect the browser performance especially for Safari on iOS.

**Comment by Zireael07** [comment]
Please make a separate issue instead of commenting on a 2 years old merged PR
